python-sdk/src/astro/files/locations/google/gdrive.py

```diff
@@ -0,0 +1,150 @@
+from __future__ import annotations
+
+import pathlib
+import urllib.parse
+from typing import Any, Iterator, Sequence
+
+from airflow.providers.google.suite.hooks.drive import GoogleDriveHook
+
+from astro.constants import FileLocation
+from astro.files.locations.base import BaseFileLocation
+
+
+def _quote(s: str) -> str:
+    s = s.replace("'", "\\'")  # Escape single quote characters.
+    return f"'{s}'"  # Quote.
+
+
+def _find_item_id(path_parts: str, *, conn: Any) -> str:
+    """Drill down the directory hierarchy to find the ID of an item.
+
+    Google Drive does not have a very good random access API by path, so we
+    must look at each directory, level by level, to find the item we want.
+    """
+    current_parent = "root"
+    folders = path_parts.split("/")
+    # First tries to enter directories
+    for current_folder in folders:
+        conditions = [
+            "mimeType = 'application/vnd.google-apps.folder'",
+            f"name='{current_folder}'",
+        ]
+        if current_parent != "root":
+            conditions.append(f"'{current_parent}' in parents")
+        result = (
+            conn.files().list(q=" and ".join(conditions), spaces="drive", fields="files(id, name)").execute()
+        )
+        files = result.get("files", [])
+        if not files:
+            # If the directory does not exist, break loops
+            break
+        current_parent = files[0].get("id")
+    return current_parent
+
+
+def _find_contains(folder_id: str, pattern: str, *, conn: Any) -> Iterator[str]:
+    """Find files in a directory matching given pattern.
+
+    This uses *contains* to search. See Google Drive documentation to
+    understand the implications:
+    https://developers.google.com/drive/api/guides/ref-search-terms
+    """
+    page_token = None
+    while True:
+        response = (
+            conn.files()
+            .list(
+                q=f"name contains {_quote(pattern)} and {_quote(folder_id)} in parents ",
+                fields="nextPageToken, files(name,id, mimeType,webContentLink,size)",
```

Comment:
What would happen when pageSize is larger than default size of 100? We should iterate through all the pages

Reply:
we do have `nextPageToken` to received in the fields to iterate over it.
https://github.com/astronomer/astro-sdk/blob/gdrive/python-sdk/src/astro/files/locations/google/gdrive.py#L64-L66 has the iterate for next pages

Reply:
Cool